File: py_src/taskito/notes.py
"""Validation + encoding for the structured ``notes`` field on jobs.

Notes are a small, bounded, user-readable annotation dict attached to a
job at enqueue time. Storage represents them as a JSON-encoded string in
the ``jobs.notes`` column across SQLite, Postgres, and Redis.

Validation is performed in Python (the single boundary that all enqueue
paths funnel through) so the Rust/PyO3 layer can stay schema-agnostic:
it accepts an already-encoded string and stores it verbatim.

Contract:

* At most :data:`MAX_NOTE_FIELDS` top-level keys.
* Each key is a non-empty ``str`` no longer than
  :data:`MAX_NOTE_KEY_LENGTH` characters.
* Nested values may be JSON-serializable primitives, lists, or dicts up
  to :data:`MAX_NOTE_DEPTH` levels of nesting.
* String leaves are capped at :data:`MAX_NOTE_VALUE_LENGTH` characters.
* The final encoded JSON document fits in :data:`MAX_NOTE_BYTES` bytes.

Limits are chosen to be tight enough that the dashboard can render the
full note set as a fixed-size key/value table without truncation.
"""

from __future__ import annotations

import json
from typing import Any

from taskito.exceptions import NotesValidationError

__all__ = [
    "MAX_NOTE_BYTES",
    "MAX_NOTE_DEPTH",
    "MAX_NOTE_FIELDS",
    "MAX_NOTE_KEY_LENGTH",
    "MAX_NOTE_VALUE_LENGTH",
    "validate_and_encode_notes",
]

MAX_NOTE_FIELDS: int = 15
MAX_NOTE_KEY_LENGTH: int = 64
MAX_NOTE_VALUE_LENGTH: int = 500
MAX_NOTE_DEPTH: int = 3
MAX_NOTE_BYTES: int = 4096

# Internal: JSON primitives accepted as leaf values.
_PRIMITIVE_TYPES = (str, int, float, bool, type(None))


def validate_and_encode_notes(notes: dict[str, Any] | None) -> str | None:
    """Validate ``notes`` and return its canonical JSON encoding.

    Returns ``None`` if ``notes`` is ``None`` (the absence of notes is
    distinct from an empty dict, which is also accepted and encoded as
    ``"{}"``).

    Raises:
        NotesValidationError: if any rule in the module docstring is
            violated. The exception message names the offending key or
            constraint so callers can surface it to end users.
    """
    if notes is None:
        return None
    if not isinstance(notes, dict):
        raise NotesValidationError(f"notes must be a dict, got {type(notes).__name__}")
    if len(notes) > MAX_NOTE_FIELDS:
        raise NotesValidationError(
            f"notes may not have more than {MAX_NOTE_FIELDS} fields, got {len(notes)}"
        )

    for key, value in notes.items():
        _validate_key(key)
        _validate_value(key, value, depth=1)

    encoded = json.dumps(notes, sort_keys=True, ensure_ascii=False, separators=(",", ":"))
    encoded_bytes = len(encoded.encode("utf-8"))
    if encoded_bytes > MAX_NOTE_BYTES:
        raise NotesValidationError(
            f"encoded notes are {encoded_bytes} bytes, exceeds limit of {MAX_NOTE_BYTES} bytes"
        )
    return encoded


def _validate_key(key: object) -> None:
    if not isinstance(key, str):
        raise NotesValidationError(f"note keys must be strings, got {type(key).__name__}")
    if not key:
        raise NotesValidationError("note keys may not be empty strings")
    if len(key) > MAX_NOTE_KEY_LENGTH:
        raise NotesValidationError(
            f"note key {key!r} is {len(key)} characters, exceeds limit of {MAX_NOTE_KEY_LENGTH}"
        )


def _validate_value(path: str, value: Any, *, depth: int) -> None:
    if depth > MAX_NOTE_DEPTH:
        raise NotesValidationError(
            f"note value at {path!r} exceeds max nesting depth of {MAX_NOTE_DEPTH}"
        )
    if isinstance(value, str):
        if len(value) > MAX_NOTE_VALUE_LENGTH:
            raise NotesValidationError(
                f"note value at {path!r} is {len(value)} characters, exceeds "
                f"limit of {MAX_NOTE_VALUE_LENGTH}"
            )
        return
    if isinstance(value, (bool, int, float)) or value is None:
        return
    if isinstance(value, list):
        for i, item in enumerate(value):
            _validate_value(f"{path}[{i}]", item, depth=depth + 1)
        return
    if isinstance(value, dict):
        for k, v in value.items():
            _validate_key(k)
            _validate_value(f"{path}.{k}", v, depth=depth + 1)
        return
    raise NotesValidationError(
        f"note value at {path!r} has unsupported type {type(value).__name__}; "
        "must be JSON-serializable (str, int, float, bool, None, list, dict)"
    )

File: sdks/java/src/main/java/org/byteveda/taskito/serialization/Notes.java
package org.byteveda.taskito.serialization;

import com.fasterxml.jackson.core.JsonProcessingException;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.databind.ObjectWriter;
import com.fasterxml.jackson.databind.SerializationFeature;
import java.nio.charset.StandardCharsets;
import java.util.List;
import java.util.Map;
import org.byteveda.taskito.errors.NotesValidationException;

/**
 * Validation and canonical encoding for the structured {@code notes} field on a job.
 *
 * <p>Notes are a small, bounded, user-readable annotation map attached at enqueue time and
 * stored as a JSON string in the {@code jobs.notes} column. Validation lives in the SDK (the
 * single boundary every enqueue funnels through) so the native layer stays schema-agnostic and
 * stores the encoded string verbatim. The bounds keep a note set renderable as a fixed-size
 * key/value table without truncation.
 *
 * <p>Contract: at most {@link #MAX_FIELDS} top-level keys; each key a non-empty string no longer
 * than {@link #MAX_KEY_LENGTH} characters; values are JSON primitives, lists, or maps nested at
 * most {@link #MAX_DEPTH} levels; string leaves at most {@link #MAX_VALUE_LENGTH} characters; and
 * the encoded document at most {@link #MAX_BYTES} UTF-8 bytes.
 */
public final class Notes {
    public static final int MAX_FIELDS = 15;
    public static final int MAX_KEY_LENGTH = 64;
    public static final int MAX_VALUE_LENGTH = 500;
    public static final int MAX_DEPTH = 3;
    public static final int MAX_BYTES = 4096;

    // Canonical form: keys sorted, compact separators (Jackson's default), non-ASCII emitted raw.
    private static final ObjectWriter CANONICAL =
            new ObjectMapper().configure(SerializationFeature.ORDER_MAP_ENTRIES_BY_KEYS, true).writer();

    private Notes() {}

    /**
     * Validate {@code notes} and return its canonical JSON encoding. A {@code null} map returns
     * {@code null} (the absence of notes, distinct from an empty map which encodes as {@code "{}"}).
     *
     * @throws NotesValidationException if any bound in the class contract is violated
     */
    public static String encode(Map<String, ?> notes) {
        if (notes == null) {
            return null;
        }
        if (notes.size() > MAX_FIELDS) {
            throw new NotesValidationException(
                    "notes may not have more than " + MAX_FIELDS + " fields, got " + notes.size());
        }
        for (Map.Entry<String, ?> entry : notes.entrySet()) {
            validateKey(entry.getKey());
            validateValue(entry.getKey(), entry.getValue(), 1);
        }
        String encoded;
        try {
            encoded = CANONICAL.writeValueAsString(notes);
        } catch (JsonProcessingException e) {
            throw new NotesValidationException("notes are not serializable: " + e.getMessage(), e);
        }
        int bytes = encoded.getBytes(StandardCharsets.UTF_8).length;
        if (bytes > MAX_BYTES) {
            throw new NotesValidationException(
                    "encoded notes are " + bytes + " bytes, exceeds limit of " + MAX_BYTES + " bytes");
        }
        return encoded;
    }

    private static void validateKey(Object key) {
        if (!(key instanceof String)) {
            throw new NotesValidationException("note keys must be strings, got " + typeName(key));
        }
        String k = (String) key;
        if (k.isEmpty()) {
            throw new NotesValidationException("note keys may not be empty strings");
        }
        if (k.length() > MAX_KEY_LENGTH) {
            throw new NotesValidationException(
                    "note key '" + k + "' is " + k.length() + " characters, exceeds limit of " + MAX_KEY_LENGTH);
        }
    }

    private static void validateValue(String path, Object value, int depth) {
        if (depth > MAX_DEPTH) {
            throw new NotesValidationException(
                    "note value at '" + path + "' exceeds max nesting depth of " + MAX_DEPTH);
        }
        if (value == null || value instanceof Boolean) {
            return;
        }
        if (value instanceof String s) {
            if (s.length() > MAX_VALUE_LENGTH) {
                throw new NotesValidationException("note value at '" + path + "' is " + s.length()
                        + " characters, exceeds limit of " + MAX_VALUE_LENGTH);
            }
            return;
        }
        if (value instanceof Double || value instanceof Float) {
            if (!Double.isFinite(((Number) value).doubleValue())) {
                // NaN/Infinity serialize to bare tokens that strict JSON readers reject.
                throw new NotesValidationException(
                        "note value at '" + path + "' must be a finite number, got " + value);
            }
            return;
        }
        if (value instanceof Number) {
            return;
        }
        if (value instanceof List<?> list) {
            for (int i = 0; i < list.size(); i++) {
                validateValue(path + "[" + i + "]", list.get(i), depth + 1);
            }
            return;
        }
        if (value instanceof Map<?, ?> map) {
            for (Map.Entry<?, ?> entry : map.entrySet()) {
                validateKey(entry.getKey());
                validateValue(path + "." + entry.getKey(), entry.getValue(), depth + 1);
            }
            return;
        }
        throw new NotesValidationException("note value at '" + path + "' has unsupported type " + typeName(value)
                + "; must be a string, number, boolean, null, list, or map");
    }

    private static String typeName(Object value) {
        return value == null ? "null" : value.getClass().getSimpleName();
    }
}
